### PR TITLE
WIP: Add process notification action and update handle notification

### PR DESCRIPTION
**Pull request by vinothvino42**
Ticket: #1166
In this PR, I updated the following

1. Created a new notification action called `ProcessNotificationAction`. Which will return the terminated notification object if it's not null or else return the `Ensemble.externalDataContext` which contains the `notificationPayload` object. User can able to get the data using notificationPayload key (it contains the title, body and data of the notification)
2. Updated the firebase message listeners (onMessage, onMessageOpened, etc) to store the payload into the `Ensemble.externalDataContext`

Sample EDL:
```yaml
View:
  styles:
    useSafeArea: true
  header:
    title: Push Notification Test
  onLoad: |
    //@code
    ensemble.storage.payloadData = "Not Found";
    if (notificationPayload != null) {
      ensemble.storage.payloadData = notificationPayload;
    }
  body:
    Column:
      styles:
        gap: 20
        padding: 20
        margin: 20
      children:
        - Button:
            label: Print Device Token
            styles:
              backgroundColor: blue
              borderRadius: 8
            onTap:
              getDeviceToken:
                onSuccess: |-
                  console.log("Device Token: "+event.data.token);

        - Divider

        - Button:
            label: Get Notification Payload
            styles:
              backgroundColor: blue
              borderRadius: 8
            onTap: |
              console.log("Get Notification");
              if (notificationPayload != null) {
                console.log(notificationPayload);
              }

        - Markdown:
            text: ${ensemble.storage.payloadData}

        - Divider
        - Button:
            label: PROCESS Notification Action
            onTap:
              processNotification:
                onNotification: |
                  console.log("processNotification - Notification Payload");
                  console.log(event.data.notificationPayload);
                onNoNotification: |
                  console.log("Failed to process notification");

```

### CODE REVIEW

**Comment by kmahmood74** [comment]
@vinothvino42 @vusters I don't get it, when would someone invoke a `processNotification` onTap of a button? I thought this action would be invoked by the platform when a notification arrives. 

It makes sense for user to register an action handler for this action outside of any screens, at the app level just like we have app settings etc.

File: lib/framework/notification_manager.dart [hunk 1]
```diff
@@ -120,17 +125,15 @@ class NotificationManager {
...
+        messageData?['screenId'] == null && messageData?['screenName'] == null;
+    ScreenController().navigateToScreen(Utils.globalAppKey.currentContext!,
         screenId: messageData!['screenId'],
         screenName: messageData['screenName'],
         pageArgs: messageData,
-      );
-    } else {
-      log('No screenId nor screenName provided on the notification. Ignoring ...');
-    }
+        routeOption: screenNotFound ? RouteOption.clearAllScreens : null);
```
Comment (vusters) [inline lib/framework/notification_manager.dart:143]: i know we talked about this but it's probably not safe to clear all the screens. It's OK to have duplicate screens

File: lib/framework/data_context.dart [hunk 2]
```diff
@@ -274,6 +274,10 @@ class DataContext {
       [SourceLocation? startLoc]) {
     try {
       _contextMap['getStringValue'] = Utils.optionalString;
+      // TODO: Added temporarily because the externalDataContext datas are not exposed when it's inside the nested callback methods
+      // Ex: Not working when this (widget.dataContext.addDataContext({'messageText': 'Data - EnsembleUI'});) inside the
+      // addPostFrameCallback method in screen.dart
+      _contextMap.addAll(Ensemble.externalDataContext);
```
Comment (vusters) [inline lib/framework/data_context.dart:280]: This is wrong. I don't understand why you need this. Also notification payload should be sent only ONCE and ONE ONLY, when the notification was delivered (or when you click on it).

File: lib/framework/notification_manager.dart [hunk 3]
```diff
@@ -124,7 +130,8 @@ class NotificationManager {
   }
 
   Future<void> _handleNotification() async {
-    Map<String, dynamic>? messageData = Ensemble.externalDataContext['data'];
+    Map<String, dynamic>? messageData =
```
Comment (vusters) [inline lib/framework/notification_manager.dart:133]: why do you have to read the data from externalDataContext? What is externalDataContext? The RemoteMessage is already there.

**Comment by vusters** [comment]
@vinothvino42 please abandon this review. There some serious issues with this when the fix is very simple. See it https://github.com/EnsembleUI/ensemble/pull/1175/commits/aa69419668a00f37e8e1ee9e100bfa7dd11f990a

**Pull request closed without merging**